src/main/cljc/jarl/builtins/units.cljc

```diff
@@ -0,0 +1,194 @@
+     "ki"   (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 1)))
+     "kib"  (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 1)))
+     "mi"   (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 2)))
+     "mib"  (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 2)))
+     "gi"   (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 3)))
+     "gib"  (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 3)))
+     "ti"   (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 4)))
+     "tib"  (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 4)))
+     "pi"   (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 5)))
+     "pib"  (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 5)))
+     "ei"   (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 6)))
+     "eib"  (bit-shift-left (js/BigInt 1) (* (js/BigInt 10) (js/BigInt 6)))
+
+     "k"    (js/BigInt 1000)
+     "kb"   (js/BigInt 1000)
+     "m"    (* (js/BigInt 1000) (js/BigInt 1000))
+     "mb"   (* (js/BigInt 1000) (js/BigInt 1000))
+     "g"    (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "gb"   (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "t"    (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "tb"   (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "p"    (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "pb"   (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "e"    (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "eb"   (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+
+     "siMilli" 0.001
+     "siK"     (js/BigInt 1000)
+     "siM"     (* (js/BigInt 1000) (js/BigInt 1000))
+     "siG"     (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "siT"     (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "siP"     (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))
+     "siE"     (* (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000) (js/BigInt 1000))}))
+
+(defn- up-unit->val [unit]
+  (case unit
+    ""          (get unit-map "")
+    "m"         (get unit-map "siMilli")
+    "M"         (get unit-map "siM")
+    ("k" "K")   (get unit-map "siK")
+    ("ki" "Ki") (get unit-map "ki")
+    ("mi" "Mi") (get unit-map "mi")
+    ("g" "G")   (get unit-map "siG")
+    ("gi" "Gi") (get unit-map "gi")
+    ("t" "T")   (get unit-map "siT")
+    ("ti" "Ti") (get unit-map "ti")
+    ("p" "P")   (get unit-map "siP")
+    ("pi" "Pi") (get unit-map "pi")
+    ("e" "E")   (get unit-map "siE")
+    ("ei" "Ei") (get unit-map "ei")
+    (throw (errors/builtin-ex "units.parse: unit %s not recognized" unit))))
```

Comment:
Would be nice if we could infer the built-in name for the exceptions. Perhaps by catching and re-throwing in the evaluator 🤔 . Are there exceptions where we don't expect the built-in name to be present?
Alternatively, we could make the built-in name an exception parameter, to make it more explicit.

Reply:
Yeah, have thought of that too. The evaluator always knows the name of the builtin getting called, so it could just prepend that to any builtin exceptions thrown. Would be nice to remove that from all the locations where thrown..